### PR TITLE
Compactor: keep bucket index updated

### LINKED ISSUE #3398: Need metric for monitoring number of blocks per tenant
Cortex should export metrics about number of blocks per each tenant, or perhaps even number of blocks per tenant per time range. Soft-deleted blocks should be counted specially.

This metric should be used for alerting if number of blocks (especially 2h blocks) per tenant is too large. Normally (if compactor isn't lagging behind) it should be between N (number of ingesters) and 0. If it grows over 2 * N  or even more, then compaction is too slow or even broken for given tenant.

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
This PR is another step forward towards the implementation of the bucket index for the blocks storage ([see proposal](https://cortexmetrics.io/docs/proposals/blocks-storage-bucket-index/)). In this PR:

- Keep bucket index updated by compactor's blocks cleaner
- Migrate block deletion marks to the per-tenant global location at compactor startup (with a flag to disable it once done)
- Added new per-tenant metrics exposed by the compactor with the number of blocks and timestamp of the last bucket index update (these metrics are removed once a tenant is deleted or moved to a different shard)

**Which issue(s) this PR fixes**:
Fixes #3398

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storage/tsdb/bucketindex/index.go [hunk 1]
```diff
@@ -33,16 +33,33 @@ type Index struct {
...
 	// List of block deletion marks.
-	BlockDeletionMarks []*BlockDeletionMark `json:"block_deletion_marks"`
+	BlockDeletionMarks BlockDeletionMarks `json:"block_deletion_marks"`
 
 	// UpdatedAt is a unix timestamp (seconds precision) of when the index has been updated
 	// (written in the storage) the last time.
 	UpdatedAt int64 `json:"updated_at"`
 }
 
+// RemoveBlock from the index.
```
Comment (pstibrany) [inline pkg/storage/tsdb/bucketindex/index.go:46]:
```suggestion
// RemoveBlock removes block and its deletion mark (if any) from index.
```

File: pkg/storage/tsdb/bucketindex/storage.go [hunk 2]
```diff
@@ -48,3 +49,40 @@ func ReadIndex(ctx context.Context, bkt objstore.Bucket, userID string, logger l
 
 	return index, nil
 }
+
+// WriteIndex generates the bucket index and writes it to the storage. If the old index is not
+// passed in input, then the bucket index will be generated from scratch.
+func WriteIndex(ctx context.Context, bkt objstore.Bucket, userID string, idx *Index) error {
```
Comment (pstibrany) [inline pkg/storage/tsdb/bucketindex/storage.go:55]: After copy, comment for this function doesn't match what it actually does.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,7 +2,14 @@
 
 ## master / unreleased
 
+* [CHANGE] Blocks storage: compactor is now required when running a Cortex cluster with the blocks storage, because it also keeps the bucket index updated. #3583
+* [CHANGE] Blocks storage: block deletion marks are now stored in a per-tenant global markers/ location too, other than within the block location. The compactor, at startup, will copy deletion marks from the block location to the global location. This migration is required only once, so you can safely disable it via `-compactor.block-deletion-marks-migration-enabled=false` once the new compact has successfully started once in your cluster. #3583
```
Comment (pstibrany) [inline CHANGELOG.md:6]: ... once new compactor has ... (instead of "compact")

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,7 +2,14 @@
...
+* [ENHANCEMENT] Blocks storage: introduced a per-tenant bucket index, periodically updated by the compactor, used to avoid full bucket scanning done by queriers and store-gateways. The bucket index is updated by the compactor every `-compactor.cleanup-interval`. #3553 #3555 #3561 #3583
```
Comment (pstibrany) [inline CHANGELOG.md:7]: > The bucket index is updated by the compactor every `-compactor.cleanup-interval`.

The flag name sounds a bit strange. I would explain it a bit:

> The bucket index is updated by the compactor during blocks cleanup, on every `-compactor.cleanup-interval`.

**Review by pstibrany — APPROVED** [review]
LGTM, great work! Pre-approving, but please check my comments before merging.

File: pkg/compactor/compactor.go [hunk 5]
```diff
@@ -78,12 +82,14 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	f.BoolVar(&cfg.ShardingEnabled, "compactor.sharding-enabled", false, "Shard tenants across multiple compactor instances. Sharding is required if you run multiple compactor instances, in order to coordinate compactions and avoid race conditions leading to the same tenant blocks simultaneously compacted by different instances.")
 	f.DurationVar(&cfg.DeletionDelay, "compactor.deletion-delay", 12*time.Hour, "Time before a block marked for deletion is deleted from bucket. "+
-		"If not 0, blocks will be marked for deletion and compactor component will delete blocks marked for deletion from the bucket. "+
-		"If delete-delay is 0, blocks will be deleted straight away. Note that deleting blocks immediately can cause query failures, "+
+		"If not 0, blocks will be marked for deletion and compactor component will permanently delete blocks marked for deletion from the bucket. "+
+		"If 0, blocks will be deleted straight away. Note that deleting blocks immediately can cause query failures, "+
 		"if store gateway still has the block loaded, or compactor is ignoring the deletion because it's compacting the block at the same time.")
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/compactor/compactor.go:91]: True, fixed.

File: pkg/compactor/blocks_cleaner.go [hunk 6]
```diff
@@ -13,22 +12,21 @@ import (
...
 type BlocksCleanerConfig struct {
-	DataDir             string
-	MetaSyncConcurrency int
-	DeletionDelay       time.Duration
-	CleanupInterval     time.Duration
-	CleanupConcurrency  int
+	DeletionDelay                      time.Duration
+	CleanupInterval                    time.Duration
+	CleanupConcurrency                 int
+	BlockDeletionMarksMigrationEnabled bool // TODO Remove in Cortex 1.8.0 and document that upgrading to 1.7.0 before 1.8.0 is required.
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/compactor/blocks_cleaner.go:29]: Relaxed the comment to a "Discuss..."

File: pkg/storage/tsdb/bucketindex/markers.go [hunk 7]
```diff
@@ -36,3 +44,47 @@ func IsBlockDeletionMarkFilename(name string) (ulid.ULID, bool) {
...
+			continue
+		} else if err != nil {
+			errs.Add(err)
+			continue
+		}
+
+		// Upload it to the global markers location.
+		uploadErr := userBucket.Upload(ctx, BlockDeletionMarkFilepath(blockID), reader)
+		if closeErr := reader.Close(); closeErr != nil {
+			errs.Add(uploadErr)
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/tsdb/bucketindex/markers.go:81]: Damn, thanks for spotting it!

File: pkg/storage/tsdb/bucketindex/storage.go [hunk 8]
```diff
@@ -48,3 +49,40 @@ func ReadIndex(ctx context.Context, bkt objstore.Bucket, userID string, logger l
...
+		return errors.Wrap(err, "upload bucket index")
+	}
+
+	return nil
+}
+
+// DeleteIndex deletes the bucket index from the storage.
+func DeleteIndex(ctx context.Context, bkt objstore.Bucket, userID string) error {
+	bkt = bucket.NewUserBucketClient(userID, bkt)
+	return errors.Wrap(bkt.Delete(ctx, IndexCompressedFilename), "delete bucket index")
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/tsdb/bucketindex/storage.go]: Very interesting, I didn't know! Turns out S3 is idempontent, but GCS is not. I've handled it.

File: pkg/storage/tsdb/bucketindex/storage.go [hunk 9]
```diff
@@ -80,8 +80,13 @@ func WriteIndex(ctx context.Context, bkt objstore.Bucket, userID string, idx *In
...
 
-// DeleteIndex deletes the bucket index from the storage.
+// DeleteIndex deletes the bucket index from the storage. No error is returned if the index
+// does not exist.
 func DeleteIndex(ctx context.Context, bkt objstore.Bucket, userID string) error {
 	bkt = bucket.NewUserBucketClient(userID, bkt)
-	return errors.Wrap(bkt.Delete(ctx, IndexCompressedFilename), "delete bucket index")
+	err := errors.Wrap(bkt.Delete(ctx, IndexCompressedFilename), "delete bucket index")
+	if err != nil && !bkt.IsObjNotFoundErr(err) {
+		return err
```
Comment (pstibrany) [inline pkg/storage/tsdb/bucketindex/storage.go:89]: nonblocking-nit: we can wrap here, instead of before `if`

**Review by pstibrany — APPROVED** [review]

File: pkg/storage/tsdb/bucketindex/storage.go (hunk 9 above)
Comment (pracucci) [reply pkg/storage/tsdb/bucketindex/storage.go]: Oh definitely. Done!

**Pull request merged**